e2e_tests/tests/experiment/test_custom_searcher.py

```diff
@@ -219,3 +225,314 @@ def sample_params(self) -> Dict[str, int]:
+    experiment_id = search_runner.run(config, context_dir=conf.fixtures_path("no_op"))
+
+    assert client._determined is not None
+    session = client._determined._session
+    response = bindings.get_GetExperiment(session, experimentId=experiment_id)
+
+    assert response.experiment.numTrials == 5
+    assert search_method.pending_trials == 0
+    assert search_method.completed_trials == 5
+    assert len(search_method.searcher_state.trials_closed) == len(search_method.closed_trials)
+
+    response_trials = bindings.get_GetExperimentTrials(session, experimentId=experiment_id)
+    for trial in response_trials.trials:
+        assert trial.state == bindings.determinedexperimentv1State.STATE_COMPLETED
+        # number of batch for rung[0] = 750
+        assert trial.totalBatchesProcessed >= 750
+
+
+@dataclass
+class TrialMetric:
+    request_id: uuid.UUID
+    metric: float
+    promoted: bool = False
+
+
+@dataclass
+class Rung:
+    units_needed: int
+    idx: int
+    metrics: List[TrialMetric] = dataclasses.field(default_factory=list)
+    outstanding_trials: int = 0
+
+    def promotions_async(
+        self, request_id: uuid.UUID, metric: float, divisor: int
+    ) -> List[uuid.UUID]:
+        logging.info(f"Rung {self.idx}")
+        logging.info(f"outstanding_trials {self.outstanding_trials}")
+
+        old_num_promote = int(len(self.metrics) / divisor)
+        num_promote = int((len(self.metrics) + 1) / divisor)
+
+        index = self._search_metric_index(metric)
+        promoteNow = index < num_promote
+        trial_metric = TrialMetric(request_id=request_id, metric=metric, promoted=promoteNow)
+        self.metrics.insert(index, trial_metric)
+
+        if promoteNow:
+            return [request_id]
+        elif num_promote != old_num_promote and not self.metrics[old_num_promote].promoted:
+            self.metrics[old_num_promote].promoted = True
+            return [self.metrics[old_num_promote].request_id]
+
+        logging.info("No promotion")
+        return []
+
+    def _search_metric_index(self, metric: float) -> int:
+        i: int = 0
+        j: int = len(self.metrics)
+        while i < j:
+            mid = (i + j) >> 1
+            if self.metrics[mid].metric <= metric:
+                i = mid + 1
+            else:
+                j = mid
+        return i
+
+
+class ASHASearchMethod(SearchMethod):
+    def __init__(
+        self,
+        max_length: int,
+        max_trials: int,
+        num_rungs: int,
+        divisor: int = 2,
+        max_concurrent_trial: int = 0,
+    ) -> None:
+        super().__init__()
+
+        # Asha params
+        self.max_length = max_length
+        self.max_trials = max_trials
+        self.num_rungs = num_rungs
+        self.divisor = divisor
+        self.max_concurrent_trials = max_concurrent_trial
+
+        self.is_smaller_better = True
+
+        # structs
+        self.rungs: List[Rung] = []
+        self.trial_rungs: Dict[uuid.UUID, int] = {}
```

Comment:
I think using a set here would result in simpler code, except for the serialization (sets are not json-serializable). Go just does not have a native set structure which is why we use maps to represent sets.

Reply:
While early_exit_trials and closed_trials should be sets in Python, trial_rungs actually maps a trial to a rung that the trial is currently in.